Removed no longer supported SQL resource types and renamed several operations

jaredmoo opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [x] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.

jaredmoo committed:
Removed upgrade APIs which are no longer useful.

These APIs were used to upgrade servers from v2 to v12. This is no longer applicable because all servers are v12.

jaredmoo committed:
Removed unsupported schemas APIs

jaredmoo committed:
Ranemd DatabaseEditions enum to DatabaseEdition.

Same for ElasticPoolEditions. Enum names should be singular.

msftclas commented:
@jaredmoo,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

jaredmoo commented:
Clarification: I actually removed inaccessible or no longer supported _types_, not APIs. The enum rename is included to get breaking changes done all together.

veronicagg reviewed (commented):
Thanks @jaredmoo Your changes look good to me, I've labeled this PR as "potential-sdk-breaking-changes".
There are other errors listed by our validation tool, not related to the changes in this PR, would you like to take a look at those as part of this PR?

jaredmoo commented:
Thanks @veronicagg , I assume you're talking about the following checks that have PR_ONLY = TRUE?
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/208461582
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/208461584

jaredmoo commented:
I believe we are waiting on you guys to make changes to the examples validation, so I will just focus on the linter errors ( https://travis-ci.org/Azure/azure-rest-api-specs/jobs/208461582 )

jaredmoo commented:
Comments on errors:
- ListByOperationsValidation: I believe the existing names are correct and we should silence these warnings somehow.
- ProvidersPathValidation: in these case `current` is a singleton, I'm interested in your guidance on this. Should we make it a {param} with only one allowed value (`current`?)
- BodyTopLevelProperties: isn't `kind` a standard property? Either way this property is what the API has which can't be changed
- ArmResourcePropertiesBag: I don't understand what this is saying. It sounds like it's complaining that we are replacing the base properties with our own properties. Can you clarify how to fix this?
- BooleanPropertyNotRecommended, GuidValidation, DefinitionsPropertiesNamesCamelCase: This is what the API does, I can't change that. We should silence these warnings.
- TrackedResourceValidation: Let me check if this is actually a tracked resource.
- TrackedResourcePatchOperationValidation: Understood, I believe patch is already supported. Will investigate.

This is the output of latest version of AutoRest that I'm looking at, right?

veronicagg commented:
@jaredmoo Regarding linter errors:

1. ListByOperationsValidation - I agree.
2. ProvidersPathValidation - Yes, so the recommendation would be to add a parameter, and make it an enum with 1 value (https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-extensions.md#single-value-enum-as-a-constant)
2. BodyTopLevelProperties -  Ok, @ravbhatnagar is kind a standard property?
3.  ArmResourcePropertiesBag - ignore this one for now, we've temporarily removed this rule, as it needs modifications
4. BooleanPropertyNotRecommended, GuidValidation, DefinitionsPropertiesNamesCamelCase - Ok, let's keep it as is for now.
5. TrackedResouceValidation - sounds good
6. TrackedResourcePatchOperationValidation - sounds good

Regarding model validation errors:
- We may have missed this error before: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/208461584#L1640, there looks to be an extra "2" in https://github.com/Azure/azure-rest-api-specs/blob/master/arm-sql/2014-04-01/examples/DatabaseList.json#L22
- For the rest, yes, we agreed on suppressing those.

Thanks!

veronicagg reviewed (changes requested):
Please take a look at my previous comment and let me know if you have additional questions. Thanks!

jaredmoo committed:
Make TDE '/current' an enum param

jaredmoo committed:
TDE config is method param

jaredmoo committed:
Reinstated System2 SLO.

I had removed this in 80ebb4c3ea6ad1319c5acbe27c977083b569f73a

jaredmoo commented:
Done. System2 is actually correct, I had removed this enum element previous because I thought it was wrong.

jaredmoo commented:
Will there be a way of suppressing these errors that are acceptable? That way in the future it will be easier to catch regressions.

anudeepsharma reviewed (commented):
LGTM

jaredmoo committed:
Renamed Servers_GetByResourceGroup to Servers_Get.

There is no other way to get a server, so the simplest name makes sense.

jaredmoo committed:
Add $expand to list databases operation

jaredmoo committed:
Make DB activity operation id format as uuid

veronicagg commented on arm-sql/2014-04-01/swagger/sql.core.json:
```diff
@@ -1330,6 +1330,9 @@
             "description": "The name of the database for which setting the transparent data encryption applies."
           },
           {
+            "$ref": "#/parameters/TransparentDataEncryptionConfigNameParameter"
```
A parameter reference like this one is missing for the "get" operation below at https://github.com/Azure/azure-rest-api-specs/pull/1005/files#diff-79c4f982670d02607125a6b94696be04R1379
See https://travis-ci.org/Azure/azure-rest-api-specs/jobs/209193853#L973

jaredmoo replied:
Fixed in 7cf6aee

veronicagg reviewed (changes requested):
Left one more comment inline, that's giving an error during semantic validation.
Regarding the "exceptions", yes, we will have a way to suppress them in the future, we're working on improving the tools to support that. 
Regarding the remaining errors:
- What did you find out on the TrackedResource checks?
- BodyTopLevelProperties  - kind is a standard property and we'll be correcting the rule

Since the latest changes introduced "importExport.json" to the PR, you may want to take a look at validation failures:
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/209193852#L964

jaredmoo committed:
Added missing TDE config name parameter

jaredmoo committed:
Merge remote-tracking branch 'azure/master' into breakingchanges

jaredmoo commented:
- TrackedResource checks: yes these resources (server, db, pool) support PATCH. We will do this in a later PR. Opened issue #1018 
- ImportExport errors: fixing those now.

jaredmoo commented:
What do I do about this one? https://travis-ci.org/Azure/azure-rest-api-specs/jobs/209193852#L980

We already have `/providers/Microsoft.Sql/operations` defined in sql.core.json.

jaredmoo committed:
Fixed validator warnings for /extensions/import api

jaredmoo committed:
Modify import operation names.

Eliminates ImportExportOperations as a top level entity.

jaredmoo committed:
Tweaked CreateImportOperation description

veronicagg commented:
@jaredmoo Don't worry about https://travis-ci.org/Azure/azure-rest-api-specs/jobs/209193852#L980, as per the other specs, adding it to sql.core is fine. We'll enhance the tool later to take composite swagger into account.

veronicagg commented:
@jaredmoo one clarification, after chatting with @ravbhatnagar a bit more, the issue around ProvidersPathValidation, which I recommended you to fix by adding a parameter with a constant value, is not actually required, you can keep it this way, or you can do it as you were doing it before, we'll be updating the validation tool for that error.

jaredmoo commented:
Ok thanks @veronicagg . In that case I'll revert the path back.

veronicagg commented:
@jaredmoo There looks to be one error from examples validation you may want to take a look https://travis-ci.org/Azure/azure-rest-api-specs/jobs/209833639#L1006 
Rest looks good to me. There's an issue with CI failing, it was supposed to be resolved today, but I see it's failing for your PR and some others, so I've pinged someone else about that.

jaredmoo commented:
Well spotted. That issue should be fixed by reverting the path change.

jaredmoo committed:
Removed extensionName param, just put import in the path

jaredmoo committed:
Removed ImportExport operation paths.

Also made various tweaks to ImportExport definitions.

jaredmoo committed:
Simplied I/E definitions names and removed unnused ListResponse

jaredmoo committed:
Removed IE operation results examples

jaredmoo commented:
SDK change: https://github.com/Azure/azure-sdk-for-net/pull/2927

veronicagg commented:
@jaredmoo thanks! couple of things I noticed:
- https://travis-ci.org/Azure/azure-rest-api-specs/jobs/210142403#L1020 not sure if you've looked at the suggestion for the name of that operation. 
- https://travis-ci.org/Azure/azure-rest-api-specs/jobs/210142407#L1026 it looks like the model there is a bit confused for ImportExportResponse, should it look something like this?
        "ImportExportResponse": {
             "allOf": [
              {
                    "$ref": "#/definitions/SqlTypedSubResource"
               }
             ],
             "properties":
                {
                     "x-ms-client-flatten": true
                     "$ref": "#/definitions/ImportExportResponseProperties"
                }
            },
            "description": "Response for Import/Export Get operation."
        },
I'm still seeing CI failures, not sure why we'll follow up again on those, sorry about the delay on these.

jaredmoo committed:
Make FirewallRules top level resource.

New policy: if it can be created/deleted, it's top leve. If it's a singleton that can only be updated (e.g. TDE configuration, audit settings, threat detection settings), then it's nested. If it's readonly, then it's nested.

jaredmoo committed:
Merge from master

jaredmoo committed:
Updated audit & TD operation name based on new policy

jaredmoo committed:
Fixed errors in resource definitions

jaredmoo committed:
Merge remote-tracking branch 'azure/master' into breakingchanges

jaredmoo commented:
- Changed operation name to CreateImportOperation
- Fixed this resource definition
- Also changed a few more operation names to standardize on naming conventions

veronicagg commented on arm-sql/2014-04-01/swagger/importExport.json:
```diff
@@ -248,11 +248,7 @@
                 },
                 "properties":
                 {
-                    "allOf": [
```
I believe allOf in line https://github.com/Azure/azure-rest-api-specs/pull/1005#issuecomment-286594317 needs to come out of "properties", so swagger doesn't think "allOf" is the name of a property.

jaredmoo commented:
I read through all the warnings and I think we're good now. Phew. Thanks for your persistence @veronicagg :)

veronicagg commented:
Np :) a few more I'm seeing though... let me know if you're planning on addressing them in this PR:
- https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211160959#L1006 (looks like the example https://github.com/Azure/azure-rest-api-specs/blob/master/arm-sql/2014-04-01/examples/DatabaseSecurityAlertGet.json has a couple of properties that are not defined in the swagger)
- similarly in this case: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211160959#L1020
- enum mismatch: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211160959#L1058, for example: https://github.com/Azure/azure-rest-api-specs/blob/master/arm-sql/2014-04-01/examples/DatabaseSecurityAlertCreateMax.json#L14
- I believe this is still showing up https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211160959#L1128 because 'allOf' should be outside the "properties" object.
-https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211160959#L2430 some extra properties here too
- and here https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211160959#L2444

jaredmoo committed:
Fixed incorrect usage of "allOf"

jaredmoo commented:
- "allOf" - fixing all of these :) thanks
- enum mismatch - discussed with @anudeepsharma  at https://github.com/Azure/azure-rest-api-specs/pull/1012#issuecomment-286307270 . The API really should be an array, but the way I specced it (modelAsString enum) at least you have programmatic access to what options are supported in the comma-separated list. If you disagree then I can change the spec.
- the API is actually returning kind, but not location. So I need to remove location from example and add kind to spec.

jaredmoo committed:
Fix security policy type definitions.

Blob auditing and threat detection policy types have kind. Thread
detection policy also has location.

jaredmoo committed:
Renamed ServerFirewallRule to FirewallRule

jaredmoo committed:
Security alert policy location isn't required

jaredmoo committed:
Audit policy also has 'kind'

jaredmoo committed:
Corrected List method names

See guidelines at https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md

jaredmoo commented:
Hi @veronicagg ,

I was reading through https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md and I found this interesting:

    * Every "Resource" Model should have "location" as a "required" property as per the mandate from ARM team; no exceptions are allowed over here.
    ....
    * All resource like models should inherit from Resource

So what about resource-like models that are untracked so they do not have required location? I made a new model named `ResourceWithOptionalLocation` which doesn't have `x-ms-azure-resource` tag, do you have guidance on what to do in this case?

jaredmoo committed:
Server_List was right the first time. Reverting

jaredmoo commented:
Ok I think the warnings are good now other than the enum issues, which from my point of view is intentional

jaredmoo commented on arm-sql/2014-04-01/swagger/sql.core.json:
```diff
@@ -1302,7 +1301,7 @@
       }
     },
-    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/databases/{databaseName}/transparentDataEncryption/current": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/databases/{databaseName}/transparentDataEncryption/{configName}": {
```
Reverting

jaredmoo committed:
Reverted TDE config to hardcoded resource name

jaredmoo committed:
Reverted TDE config to hardcoded resource name (part 2)

jaredmoo commented:
I just noticed that TransparentDataEncryption config looks like a long-running operation. Need to write SDK test to verify.

cormacpayne commented:
Swagger used to generate https://github.com/Azure/azure-sdk-for-net/pull/2927

veronicagg commented:
@jaredmoo Regarding the "Resource" model, do you actually need location for anything? The basic definition of resource, should really just include "id", "name" and "type", if you're adding a *required* location, then it becomes a "tracked resource", you can take a look at the common definitions which are being reviewed at https://github.com/Azure/azure-rest-api-specs-pr/pull/40/files#diff-d3a0dcdc5d2939703fa9e32dff2295c1R22 When that PR is merged, there will be a request for teams to move to refer to this definitions, so the easier you can make it in your end to later on make the update the better. If you don't need location, don't add it, and use "x-ms-azure-resource" if the model is a "Resource" (the rule you pointed out in the "good patterns" page, may need update now that we have clarified this definitions, I'll take a look into that).
I hope that makes sense, feel free to let me know if you have more questions on this.

Thanks for updating the rest. I'm following up on the enum mismatch, to see if we have a better way of spec'ing it.

jaredmoo committed:
Renamed TransparentDataEncryptionStates

jaredmoo commented:
Thanks for the resource clarification. I'll add the definitions from Azure/azure-rest-api-specs-pr#40

jaredmoo committed:
Use common resource definitions

jaredmoo committed:
Added missing location props

jaredmoo commented:
We have some proxy resources that don't have a type property. IMO this is an API bug, but for Swagger purposes right now should I have them include `ProxyResource`, or make my own model type e.g. `SubResource` which has just id and name?

jaredmoo committed:
Add another missing location

veronicagg commented on arm-sql/2014-04-01/swagger/resourceDefinitions.json:
```diff
@@ -0,0 +1,271 @@
+{  
+  "definitions":{  
+    "Resource":{  
```
missing   "x-ms-azure-resource": true

jaredmoo replied:
Done

veronicagg commented on arm-sql/2014-04-01/swagger/resourceDefinitions.json:
```diff
@@ -0,0 +1,271 @@
+        }
+      ]
+    },
+    "ResourceModelWithAllowedPropertySet":{  
```
missing "x-ms-azure-resource": true?

jaredmoo replied:
Removed this model (unused)

veronicagg commented on arm-sql/2015-05-01-preview/swagger/resourceDefinitions.json:
```diff
@@ -0,0 +1,271 @@
+      ]
+    }
+  },
+  "parameters":{  
```
are you referring to these parameters anywhere?

jaredmoo replied:
No, removed

jaredmoo replied:
I was just copying resourceDefinitions.json from Azure/azure-rest-api-specs-pr#40 since I thought that's what you were asking for.

veronicagg commented on arm-sql/2014-04-01/swagger/resourceDefinitions.json:
```diff
@@ -0,0 +1,271 @@
+{  
```
you could include these definitions in one of your existing files and have others refer to them, having a separate file will, at this time, produce "many" violations, since this swagger file is not following an RP structure. At some point, our tools would account for this separate files.

jaredmoo replied:
Ok. I misunderstood what you said earlier. Done.

veronicagg reviewed (changes requested):
Regarding your question on "ProxyResource" versus "SubResource", since you're referencing your own definitions right now, you can create your own. 
Regarding the enum mismatch from a previous comment, I'd recommend updating the spec to make it a "string" and include in the description what the possible values are and how to construct the string, not sure if they are comma separated or semicolon separated, and also, is there a semicolon at the end? When the API bug gets fixed, then this could be turned into an array type.

jaredmoo committed:
Remove common Resource definition files.

Resource is back to being defined in each spec file.

Also improved some operation descriptions.

jaredmoo committed:
Change disabledAlerts back to string, not enum.

jaredmoo committed:
Fixed body parameters name in examples

jaredmoo commented:
Changed that enum to a regular string. It's semicolon separated, description was wrong.

jaredmoo committed:
storageAccountAccessKey is write-only

jaredmoo committed:
Fixed more incorrectly named enums

veronicagg commented:
Thanks @jaredmoo . 
Regarding the resource definitions, if they are shared among your own files, you could have them just in one of those, and refer to the ones in that one from the others. It's also fine if you want to keep it this way.
From the latest validations, have you looked into the following?
- https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211678413#L935 
- https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211678413#L1060
- https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211678413#L1239
- https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211678413#L1250
- https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211678413#L1272

jaredmoo committed:
Merge remote-tracking branch 'azure/master' into breakingchanges

jaredmoo committed:
Capabilities_List -> ListByLocation in description

jaredmoo commented:
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211678413#L935 - false positive. DatabaseSecurityAlertPolicy has location, but it's not required, so it's not a tracked resource.

    https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211678413#L1060 - unfortunately the API already uses guid in production, I can't change it

    https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211678413#L1239 - issue #1018 

Resolving merge conflicts and removing resource definitions from all files except sql.core
    https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211678413#L1250 - issue #1018 

    https://travis-ci.org/Azure/azure-rest-api-specs/jobs/211678413#L1272 - unfortunately the API already uses guid in production, I can't change it

jaredmoo committed:
Remove repeated Resource definitions.

Refer to sql.core.json definitions from other files.

jaredmoo committed:
Reverted resource property descriptions changes.

jaredmoo committed:
Removed 'DataWarehouse' from Pause/Resume op ids.

veronicagg reviewed (approved):
Changes look good, thanks!

AutorestCI commented:
No modification for NodeJS

AutorestCI commented:
[![Build Status](https://travis-ci.org/AutorestCI/azure-sdk-for-ruby.svg?branch=RestAPI-PR1005)](https://travis-ci.org/AutorestCI/azure-sdk-for-ruby) https://github.com/Azure/azure-sdk-for-ruby/pull/721

AutorestCI commented:
[![Build Status](https://travis-ci.org/AutorestCI/azure-sdk-for-python.svg?branch=RestAPI-PR1005)](https://travis-ci.org/AutorestCI/azure-sdk-for-python) https://github.com/Azure/azure-sdk-for-python/pull/1039
